### PR TITLE
[REVIEW] Parquet Reader: Fix crash due to clash with ORC reader datasource

### LINKED ISSUE #1572: [BUG] Segfault in Parquet reader when running Parquet pytests
**Describe the bug**
When building a debug build of libcudf locally and running the Parquet pytests, there is a segfault when attempting to open the input source.

**Steps/Code to reproduce bug**
1. Build cudf locally, with CMAKE_BUILD_TYPE=Debug, along with CFFI and Cython bindings
2. Run Parquet pytests
3. Observe segfault when running the first test that uses the `cudf` backend

**Expected behavior**
Parquet tests all run and pass successfully.

**Environment details (please complete the following information):**
 - Environment location: Bare-metal
 - Method of cuDF install: from source
 
**Additional context**
The root cause is due to the ORC reader having the same name of DataSource class as the Parquet reader, but differing implementation.

### PR DESCRIPTION
**Pull request by j-ieong**
Fixes #1572. 

This tries to be a minimal change. More substantial consolidation to occurr in `0.8`.

* Move `DataSource` class to separate header and share it between Parquet and ORC
    - add a cudaSynchronize() in ORC to ensure that the read is finished during buffer lifetime, like Parquet
* Move metadata log prints to separate function
* Move doxygen for `read_orc` to interface header

### CODE REVIEW

**Comment by j-ieong** [comment]
rerun tests

File: cpp/src/io/orc/orc_reader.cu [hunk 1]
```diff
@@ -711,8 +682,10 @@ gdf_error read_orc(orc_read_arg *args) {
           len += stream_info[stream_count].length;
           stream_count++;
         }
-        CUDA_TRY(cudaMemcpyAsync(d_dst, input.data() + offset, len,
+        const auto buffer = input.get_buffer(offset, len);
+        CUDA_TRY(cudaMemcpyAsync(d_dst, buffer->data(), len,
                                  cudaMemcpyHostToDevice));
+        CUDA_TRY(cudaStreamSynchronize(0));
```
Comment (vuule) [inline cpp/src/io/orc/orc_reader.cu]: I don't think this change is mentioned in the description

File: cpp/src/io/parquet/parquet_reader.cu [hunk 2]
```diff
@@ -600,26 +549,25 @@ void decode_page_data(
...
+      return DataSource(args->source);
+    } else if (args->source_type == ARROW_RANDOM_ACCESS_FILE) {
+      CUDF_EXPECTS(rd_file, "Invalid Arrow file handle");
+      return DataSource(rd_file);
+    } else {
+      return DataSource(args->source, args->buffer_size);
+    }
+  };
 
+  DataSource input = createDataSource();
```
Comment (vuule) [inline cpp/src/io/parquet/parquet_reader.cu]: Not sure if you would consider this more or less readable, but you can declare and call the initialization lambda inline with DataSource declaration. Something like this: https://www.bfilipek.com/2016/11/iife-for-complex-initialization.html

File: cpp/src/io/orc/orc_reader.cu (hunk 1 above)
Comment (j-ieong) [reply cpp/src/io/orc/orc_reader.cu]: Added. Right now, it's to ensure it matches Parquet with BlazingDB's change in that PR. May not be needed depending on how the Arrow zero-copy is implemented but we'll have to revisit.

File: cpp/src/io/parquet/parquet_reader.cu (hunk 2 above)
Comment (j-ieong) [reply cpp/src/io/parquet/parquet_reader.cu]: Cool; changed it to save a couple of lines thanks!

**Pull request merged**